### PR TITLE
[Feature] support single file external table

### PR DESCRIPTION
**Pull request by zombee0**
Signed-off-by: zombee0 <flylucas_10@163.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

`CREATE EXTERNAL TABLE `hive_0` (
  `c1` int NULL, 
  `c2` array<array<int>> NULL
) ENGINE=FILE
PROPERTIES (
  "path"="hdfs://172.17.0.3/user/hive/warehouse/array2d_parq",
  "format"="parquet"
);`

the path can be a dir or a file,
the table name and type must be same with the file in the directory or the file itself.
support orc and parquet.

then can query the hive_0:
`select * from hive_0`

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 1]
```diff
@@ -223,21 +229,63 @@ public void initHmsTableInfo() {
...
 
     @Override
     public List<HivePartition> getPartitions(List<PartitionKey> partitionKeys) throws DdlException {
+        if (singleFile) {
+            String path = hiveProperties.get(JSON_KEY_HDFS_PATH);
+
+            if (!path.isEmpty()) {
+                StringBuilder sBuilder = new StringBuilder();
+                RemoteFileInputFormat format = RemoteFileInputFormat.UNKNOWN;
+                if (hiveProperties.get(JSON_KEY_FORMAT).equalsIgnoreCase("parquet")) {
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:248]: better put into a method.

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 2]
```diff
@@ -223,21 +229,63 @@ public void initHmsTableInfo() {
...
+            return new HiveTableStats(-1, -1);
+        }
         return HiveMetaStoreTableUtils.getTableStats(hmsTableInfo);
     }
 
     @Override
     public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> columnNames) throws DdlException {
+        if (singleFile) {
+            Map<String, HiveColumnStats> result = Maps.newHashMap();
+            for (String name : dataColumnNames) {
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:284]: also better put into a single method.

**Review by dirtysalt — COMMENTED** [review]
LGTM.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 8 / 60 (13.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/HiveTable.java|7|54|12.96%|[233, 234, 235, 243, 245, 246, 247, 248, 249, 250, 251, 253, 255, 257, 259, 260, 261, 262, 263, 266, 275, 283, 284, 285, 286, 287, 404, 420, 421, 422, 423, 424, 425, 426, 428, 429, 430, 432, 433, 434, 435, 436, 437, 438, 542, 683, 684]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|1|6|16.67%|[123, 124, 125, 126, 767]|

**Comment by imay** [comment]
@zombee0 I think engine couldn't be hive, because `hive` stands for Hive table already, and this is a single file, not a hive table.

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 3]
```diff
@@ -223,21 +229,63 @@ public void initHmsTableInfo() {
     }
 
     public Map<PartitionKey, Long> getPartitionKeys() throws DdlException {
+        if (singleFile) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:232]: I’ll remove this function this week. plz don't merge it :)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 86 (24.42%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/HiveTable.java|9|65|13.85%|[204, 243, 249, 250, 251, 255, 256, 257, 258, 259, 260, 261, 263, 265, 267, 269, 270, 271, 272, 273, 276, 283, 291, 299, 300, 301, 302, 303, 310, 311, 408, 409, 416, 417, 429, 447, 448, 449, 450, 452, 453, 454, 456, 457, 458, 460, 461, 462, 463, 464, 465, 466, 570, 710, 711, 712]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|1|6|16.67%|[123, 124, 125, 126, 767]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|7|11|63.64%|[2230, 2231, 2232, 2233]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateTableStmt.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 4]
```diff
@@ -115,6 +125,7 @@ public class HiveTable extends Table implements HiveMetaStoreTable {
     private HiveMetaStoreTableInfo hmsTableInfo;
 
     private HiveRepository hiveRepository;
+    private boolean fileEngine;
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:128]: Do you think it's better to extract a `FileTable`

**Review by DorianZheng — APPROVED** [review]
left a comment, reset LGTM

**Comment by dulong41** [comment]
run starrocks_admit_test

**Pull request closed without merging**